### PR TITLE
Improves typings of .on('connection') callback - add upgrade request

### PR DESCRIPTION
**Pull request by trebler**
Related to #11 .

This fix allows for following syntax in TS:
```typescript
wsServer.on('connection', (ws, req) => { 
    ...
}
```
which would not work previously (`Parameter 'ws' implicitly has an 'any' type` and `Parameter 'req' implicitly has an 'any' type`).

### CODE REVIEW

File: lib/cws/server.ts [hunk 1]
```diff
@@ -38,7 +38,7 @@ export class WebSocketServer extends EventEmitterServer {
   // overload on function from super class
   public on(event: string, listener: Listener): void;
   public on(event: 'error', listener: (err: Error, socket?: Socket) => void): void;
-  public on(event: 'connection', listener: (socket: WebSocket) => void): void;
+  public on(event: 'connection', listener: (socket: WebSocket, upgradeRequest: HTTP.IncomingMessage) => void): void;
```
Comment (goriunov) [inline lib/cws/server.ts]: Can you make `upgradeRequest` optional.

Comment (trebler) [reply lib/cws/server.ts]: I could of course. But the only place where `connection` is emitted does include reference to upgrade request:
```typescript
this.emit('connection', webSocket, this.upgradeReq);
```
And if a user needs to use that upgrade request in the `.on('connection')` callback (e.g., I need query parameters from url), he would have to write something like:
```typescript
// req: IncomingMessage | undefined
wsServer.on('connection', (ws, req) => {
    if (!req) {
        // oops, can't continue
        ws.close();
        
        return;
    }
    
    ws.on('message', msg => { });
    // ...
});
```

Comment (goriunov) [reply lib/cws/server.ts]: Oh, yes you are right. Can you instead add additional overload for that function: 

```ts
public on(event: 'connection', listener: (socket: WebSocket) => void): void;
public on(event: 'connection', listener: (socket: WebSocket, upgradeRequest: HTTP.IncomingMessage) => void): void;
```

Comment (trebler) [reply lib/cws/server.ts]: TSLint is complaining: 
```
This overload and the one on line 41 can be combined into one signature taking `(socket: WebSocket, upgradeRequest: HTTP.IncomingMessage) => void | (socket: WebSocket) => void`. (unified-signatures)
```
Would you prefer this:
``` typescript
public on(event: 'connection', listener: (socket: WebSocket, upgradeRequest: HTTP.IncomingMessage) => void): void;
// tslint:disable-next-line: unified-signatures
public on(event: 'connection', listener: (socket: WebSocket) => void): void;
```
this:
``` typescript
public on(event: 'connection', listener: ((socket: WebSocket, upgradeRequest: HTTP.IncomingMessage) => void) | ((socket: WebSocket) => void)): void;
```
or this :):
``` typescript
public on(
    event: 'connection',
    listener:
      ((socket: WebSocket, upgradeRequest: HTTP.IncomingMessage) => void)
      | ((socket: WebSocket) => void)
  ): void;
```
NB: order of overloads needs to be opposite from what you proposed

Comment (goriunov) [reply lib/cws/server.ts]: @trebler to fix above issue can you add  rule to tslint.json ` "unified-signatures": false, ` for all new projects i have this rule in place. Thank you :)

Comment (trebler) [reply lib/cws/server.ts]: @goriunov, all done now: added overload and disabled TSLint `unified-signatures` rule.

**Comment by goriunov** [comment]
Looks good, thanks :)

**Comment by trebler** [comment]
@goriunov, I see that the order of `.on('connection')` signatures has again been swapped in commit 69eb5b40a2f95e7b5d8489fdfee84709b427478e, and this again breaks typings.

Any specific reason why you had to do that?

**Comment by goriunov** [comment]
@trebler ohhh...  looks like during refactoring missed that. Thanks for pointing that out, i have release version 1.1.2 with fix

**Pull request merged**